### ISSUE TITLE
How can I cause selected data grid rows to always be displayed even when they don't match the filters?

### ISSUE THREAD

**Issue by ryancogswell** [mui/mui-x#3895]
### Order ID 💳

36855

### Duplicates

- [X] I have searched the existing issues

### Latest version

- [X] I have tested the latest version

### The problem in depth 🔍

We are leveraging DataGridPro using checkbox selection along with filtering.

There are three main options (that I can think of) for how the checkbox/row selection and filtering could interact:

1. Rows remain selected even when excluded by filters (default behavior)
2. Automatically deselect rows excluded by filters
3. Always show selected rows even when they do not match the current filtering

Outside of the data grid, we have an "Actions" menu providing actions that the user can take on the selected rows. To avoid user confusion about which rows they are acting on, our preferred behavior would be option 3 above. Is there a way to achieve this?

### Your environment 🌎

<details>
  <summary>`npx @mui/envinfo`</summary>

```
  Don't forget to mention which browser you used.
  Output from `npx @mui/envinfo` goes here.
```
</details>

**Comment by cherniavskii** [comment]
Hey @ryancogswell 
Thanks for sharing your thoughts!
It looks like right now the only option available is nr. 1 (the default one).

I can see how option 1 is confusing when having an action for selected rows.
But option 3 might be confusing as well, since it will make filtering misleading.
I feel like option 2 is the least confusing solution here.

We'll need to explore this topic and see what data grid can offer.
Also, it's kind of related to https://github.com/mui/mui-x/issues/1141

**Comment by flaviendelangle** [comment]
This depends a lot on the use-case of the user

I'm not sure we should touch the current behavior.
Users can control de models and customize the behaviors.
Option 2 is easy to do 
Option 3 is more tricky but totally doable (see [this example](https://codesandbox.io/s/datagridpro-v5-quick-start-forked-qc1mj?file=/src/App.tsx))

**Comment by cherniavskii** [comment]
@flaviendelangle nice!
Option 3 is really tricky, especially with different column types.
But still great to see it's doable!

**Comment by flaviendelangle** [comment]
Yes it requires to override each column filter operators.
We could add a prop `skipFilteringForSelectedColumns` or something like that if we feel that it is a recurrent need. 
But I would wait :+1:

**Comment by ryancogswell** [comment]
> I feel like option 2 is the least confusing solution here.

@cherniavskii I found discussion in other issues about option 1 vs 2 and I think the current default behavior is reasonable. Option 2 is less flexible than 1 or 3 since it doesn't allow for the user to apply one filter, select all, apply different filter, select all, then act on the accumulated selections.

> But option 3 might be confusing as well, since it will make filtering misleading.

Though Option 3 could be confusing to some users, it isn't difficult to make the behavior less confusing either via a footnote or by making it a user-selectable option within the filter dialog (e.g. an "Always show selected rows" checkbox).

@flaviendelangle Thanks for the example! I did some digging through the code yesterday, but I think it would have taken me multiple days to figure out the different pieces I needed to leverage. I figured from the digging that I did, that it would need to be handled at the operator level, so your example wasn't any more complicated than what I was anticipating. I made a [modified version of your example](https://codesandbox.io/s/datagridpro-customized-operators-to-include-selected-8mj8q?file=/src/CustomGridFilterPanel.tsx) to add a checkbox to the filter panel to control the behavior. In order to get the checkbox change to have an immediate effect, I added `apiRef.current.upsertFilterItem(filterModel.items[0]);` within the checkbox `onChange` -- otherwise it didn't take effect until I changed/added/removed a filter or until I changed a selection. Is there a better way to force the filters to be reevaluated? I tried `forceUpdate`, but that didn't do the trick.

**Comment by flaviendelangle** [comment]
>  I figured from the digging that I did, that it would need to be handled at the operator level, so your example wasn't any more complicated than what I was anticipating

I woudl be interested in your feedback about the [doc of the operators](https://mui.com/components/data-grid/filtering/#customize-the-operators) and how it helped you build your example or failed to help you.


> Is there a better way to force the filters to be reevaluated?

`apiRef.current.unstable_applyFilters()`

We made this method private because we did not find any suitable scenarios where it was needed by the user.
But this issue is a counter example.
In my example I did not use it because I prefer to stay with public API.

**Comment by ryancogswell** [comment]
> I would be interested in your feedback about the doc of the operators

@flaviendelangle I had glanced at that part of the documentation before starting to dig through the code, but at that point I was still looking for something to impact the behavior at a higher level and didn't think that a custom operator was what I wanted since I wanted to change the behavior of all of the operators.

Now after seeing your example and looking through that part of the documentation, I can see that a lot of the pieces needed are documented there. The [Edit an operator](https://mui.com/components/data-grid/filtering/#edit-an-operator) example shows modifying all of the operators for a given data type. [Create a custom operator](https://mui.com/components/data-grid/filtering/#create-a-custom-operator) explains defining `getApplyFilterFn`, though without seeing an example, not sure if it would have occurred to me to be able to delegate to the default `getApplyFilterFn` while adding custom logic around it in order to transform all the operators. Even if I had figured that part out, I think I still would have struggled with the `selectionModelLookup` part of your example -- I'm not finding anything in the documentation covering that part.

So, in summary, an example that customized `getApplyFilterFn` by wrapping the original for an operator would help and an example (perhaps in the "Selection" page) doing logic based on whether or not a row is selected leveraging `models.selectionModel` similar to your example would have helped. And of course an example putting all of this together (like the one you provided here) would have helped. Though I have deep knowledge of most of MUI, I'm new to the data grid and it has a **lot** more complexity than any of the core components, so I'm still coming up to speed on the code approaches used within it. Thanks for your help!

**Comment by ryancogswell** [comment]
@flaviendelangle I had some difficulty getting the timing of `unstable_applyFilters` to work ([example](https://codesandbox.io/s/datagridpro-customized-operators-to-include-selected-forked-dr0tb?file=/src/CustomGridFilterPanel.tsx:673-694)). I had to wrap it in a `setTimeout` to get it to apply the updated filters rather than the previous filters (whereas calling `upsertFilterItem` worked without any special handling of the timing), so if you were to make `unstable_applyFilters` (or something similar, since you won't want to add extra overhead to the current private method) part of the public API, I would recommend doing a version that uses new filter definitions as reliably as what happens when calling `upsertFilterItem`.

**Comment by m4theushw** [comment]
For option 2 we could make it configurable through a new prop.

For option 3 we also need to consider server-side pagination. In this case the `rows` prop might not contain the rows that were selected in the previous page.

**Comment by flaviendelangle** [comment]
@ryancogswell thanks for the feedback on the doc

`upsertFilterItem` calls `unstable_applyFilters` behind the scene. 
Anytime the filter model is updated, your filters are re-applied (same when the rows are updated).
You only have issues if you want to re-apply filters after a modification of another substate we did not linked to the filtering.
But it is clearly not something we anticipated, so right now I think the `setTimeout` remains your best option. Or in a `useEffect` listening to `alwaysShowSelected`.

**Comment by ryancogswell** [comment]
> Or in a `useEffect` listening to `alwaysShowSelected`.

@flaviendelangle I had tried calling `unstable_applyFilters` from a `useEffect` prior to trying the `setTimeout` and was surprised that it didn't work. I suspect success is dependent on `unstable_applyFilters` being called after some effects within the data grid execute. I had initially put the `useEffect` within my `CustomGridFilterPanel` (which didn't work). If I [move the useEffect fully outside the data grid](https://codesandbox.io/s/datagridpro-customized-operators-to-include-selected-forked-etxze?file=/src/App.tsx:872-978) (e.g. in App.tsx in my sandbox) then it works fine.

**Comment by whitehat101** [comment]
I was about to request this feature (Option 3 "Always show selected rows even when they do not match the current filtering")

However, [Row Pinning](https://mui.com/x/react-data-grid/row-pinning/) is actually better for my use case. (A user is adding a new row to the grid, and needs to edit it before saving.) Pinning ignores filters and sorting. As long as you don't need the rows in the middle of the grid, pinning is a simpler and better supported solution.
